sandbox/cgroup/cgroup.go

```diff
@@ -259,3 +257,96 @@ func MockVersion(mockVersion int, mockErr error) (restore func()) {
 		probeVersion, probeErr = oldVersion, oldErr
 	}
 }
+
+// securityTagFromCgroupPath returns a security tag from cgroup path.
+func securityTagFromCgroupPath(path string) (securityTag string) {
+	leaf := filepath.Base(filepath.Clean(path))
+	// NOTE: The things we are returning need to minimally look like security
+	// tags. They are further refined and looked at below, in PidsOfSnap.
+	if matched, _ := filepath.Match("snap.*.service", leaf); matched {
+		return strings.TrimSuffix(leaf, ".service")
+	}
+	if matched, _ := filepath.Match("snap.*.scope", leaf); matched {
+		// Neither dot1 nor dot2 can be -1 because the match guarantees that at
+		// least two dots exist.
+		dot1 := strings.IndexRune(leaf, '.')
+		dot2 := strings.IndexRune(leaf[dot1+1:], '.')
+		maybeTag := strings.TrimSuffix(leaf[:dot1]+leaf[dot1+dot2+1:], ".scope")
+		if naming.ValidateSecurityTag(maybeTag) == nil {
+			return maybeTag
+		}
+	}
+	return ""
+}
+
+// PidsOfSnap returns the association of security tags to PIDs.
+//
+// NOTE: This function returns non-empty result only if refresh-app-awareness
+// is enabled.
+//
+// The return value is a snapshot of the pids of a given snap, grouped by
+// security tag. The result may be immediately stale as processes fork and
+// exit but it has the following guarantee.
+//
+// If the per-snap lock is held while computing the set, then the following
+// guarantee is true: If a security tag is not among the result then no such
+// tag can come into existence while the lock is held.
+//
+// This can be used to classify the activity of a given snap into activity
+// classes, based on the nature of the security tags encountered.
+func PidsOfSnap(snapInstanceName string) (map[string][]int, error) {
+	// pidsByTag maps security tag to a list of pids.
+	pidsByTag := make(map[string][]int)
+	securityTagPrefix := "snap." + snapInstanceName + "."
+
+	// Walk the cgroup tree and look for "cgroup.procs" files. Having found one
+	// we try to derive the snap security tag from one. If successful and the
+	// tag matches the snap we are interested in we harvest the snapshot of
+	// PIDs that belong to the cgroup and bin them into a bucket associated
+	// with the security tag.
+	walkFunc := func(path string, fileInfo os.FileInfo, err error) error {
+		if err != nil || fileInfo.IsDir() {
```

Comment:
I believe so but this is somewhat subtle on how the walker is implemented in go. We may be handed an error straight away, even before we attempt to do anything. The documentation, IIRC, allows the function to decide what to do in such case. We may also be handed a directory with we are not interested in. I will split this so that the error check is handled explicitly and first (like now) and only after that we check for directory vs other kind of file. The behaviour won't change but it will be easier to follow as to what is going on.

Reply:
Done.